pkg/volumes/alicloud/discovery.go

```diff
@@ -0,0 +1,92 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package alicloud
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/aliyun/alibaba-cloud-sdk-go/sdk/requests"
+	"github.com/aliyun/alibaba-cloud-sdk-go/services/ecs"
+	"github.com/golang/glog"
+	"kope.io/etcd-manager/pkg/privateapi/discovery"
+	"kope.io/etcd-manager/pkg/volumes"
+)
+
+// AlicloudVolumes allows us to discover our peer nodes
+var _ discovery.Interface = &AlicloudVolumes{}
+
+func (a *AlicloudVolumes) Poll() (map[string]discovery.Node, error) {
+	nodes := make(map[string]discovery.Node)
+
+	allVolumes, err := a.findVolumes(false)
+	if err != nil {
+		return nil, err
+	}
+
+	instanceToVolumeMap := make(map[string]*volumes.Volume)
+	for _, v := range allVolumes {
+		if v.AttachedTo != "" {
+			instanceToVolumeMap[v.AttachedTo] = v
+		}
+	}
+
+	if len(instanceToVolumeMap) != 0 {
+		request := ecs.CreateDescribeInstancesRequest()
+		request.RegionId = a.region
+		request.PageSize = requests.NewInteger(PageSizeLarge)
+
+		var ids []string
+		for id := range instanceToVolumeMap {
+			ids = append(ids, id)
+		}
+		request.InstanceIds = "[\"" + strings.Join(ids, "\",\"") + "\"]"
```

Comment:
looks quite dirty :) instanceids takes input as string but for real its array?

Reply:
You are right, it is a "Array string", very annoying. Please refer to the API doc [here](https://www.alibabacloud.com/help/doc-detail/25514.htm?spm=a2c63.p38356.b99.566.7b152f138kMSjB), try to search for "The IDs of the disks." in your browser, it will get to the right point. Alicloud may improve their SDK in the future, but for now, we have to do it this way. I can separate this ugly code into a small function, this might looks better.

Reply:
Added a `arrayString` func, and rebased.